Add proofs of custody (#145)
ethereum/eth2.0-specs: specs/core/1_shard-data-chains.md

|------------------------|-----------------|-------|---------------|

> | `CHUNK_SIZE`           | 2**8 (= 256)    | bytes |               |

Reviewer: ```suggestion
| `SHARD_CHUNK_SIZE` | 2**8 (= 256)    | bytes |               |
```

### Shard block fork choice rule

The fork choice rule for any shard is LMD GHOST using the validators currently assigned to that shard, but instead of being rooted in the genesis it is rooted in the block referenced in the most recent accepted crosslink (ie. `state.crosslinks[shard].shard_block_hash`). Only blocks whose `beacon_chain_ref` is the block in the main beacon chain at the specified `slot` should be considered (if the beacon chain skips a slot, then the block at that slot is considered to be the block in the beacon chain at the highest slot lower than a slot).

## Changes to beacon chain

Change to attestation verification:

> * Let `expected_depth = log2(SHARD_BLOCK_SIZE // SHARD_CHUNK_SIZE * next_power_of_2(slot - last_crosslink_slot))`. For each integer `i = 0, 1`, let `messages[i]` be `AttestationSignedData(slot, shard, parent_hashes, shard_block_hash, i == 1, expected_depth, attestation_indices.total_validator_count, justified_slot)`.

Reviewer: 1. `AttestationSignedData` was renamed to `AttestationData`, and some fields were updated.
2.  What's the `i == 1` here?

A `ProofOfCustodyChallenge` has the following fields:

```python
{
    # Which validator is responding
    'responder_index': 'uint64',
    # Seed hash
    'seed_hash': 'hash32',
    # Depth
    'depth': 'uint64',
    # Index in tree
    'data_index': 'uint64',
    # Expiry date
    'expiry_slot': 'uint64',
    # Who is challenging
    'challenger_index': 'uint64',
    # Data root
    'data_root': 'hash32',
    # Proof of custody bit
    'bit': 'bool'
}
```

Three new types of `SpecialObject` records:

#### PROOF_OF_CUSTODY_SEED_CHANGE

```python
{
    'index': 'uint64',
    'new_commitment': 'hash32',
    'signature': ['uint256']
}
```

> Let `signed_data = bytes8(fork_version) + new_commitment`. Verify that `BLSVerify(pub=state.validators[index].pubkey, msg=hash(signed_data), sig=signature)` passes. Verify that `hash(new_commitment) = state.validators[index].proof_of_custody_commitment`, and the `block.slot >= state.validators[index].proof_of_custody_last_change + PROOF_OF_CUSTODY_MIN_CHANGE_PERIOD`. Set `state.validators[index].proof_of_custody_second_last_change = state.validators[index].proof_of_custody_last_change` and `state.validators[index].proof_of_custody_last_change = block.slot`.

Reviewer: A little Justinize:
```suggestion
* Let `signed_data` be `bytes8(fork_version) + new_commitment`.
* Verify that `BLSVerify(pub=state.validators[index].pubkey, msg=hash(signed_data), sig=signature)` passes.
* Verify that `hash(new_commitment) = state.validators[index].proof_of_custody_commitment`.
* Verify that `block.slot >= state.validators[index].proof_of_custody_last_change + PROOF_OF_CUSTODY_MIN_CHANGE_PERIOD`.
* Set `state.validators[index].proof_of_custody_second_last_change` to `state.validators[index].proof_of_custody_last_change`.
* Set `state.validators[index].proof_of_custody_last_change` to `block.slot`.
```

#### PROOF_OF_CUSTODY_CHALLENGE

```
{
    'attestation': SpecialAttestationData,
    'validator_index': 'uint64',
    'data_index': 'uint64',
}
```

Perform the following checks:

* Verify that the attestation is valid.
* Verify that `attestation.slot > state.validators[index].proof_of_custody_second_last_change`.
* Verify that `validator_index in attestation.aggregate_sig_poc_0_indices` or `validator_index in attestation.aggregate_sig_poc_1_indices`; if the former, let `bit = False`, else `bit = True`.
* Verify that `state.validators[index].status == ACTIVE` or `state.validators[index].status == PENDING_EXIT` and `block.slot < state.validators[index].exit_slot + PROOF_OF_CUSTODY_MIN_CHANGE_PERIOD`
* Let `check_value = as_uint256(hash(as_bytes8(validator_index) + as_bytes8(attestation.slot) + as_bytes8(attestation.shard_id)))`.
* Let `FULL_PROOF_OF_CUSTODY_VALCOUNT = TARGET_COMMITEE_SIZE * SHARD_COUNT`, and `target = 2**256 * attestation.total_validator_count // FULL_PROOF_OF_CUSTODY_VALCOUNT`.
* Verify that `check_value <= target`.

> Let `seed_hash = state.validators[index].proof_of_custody_commitment if attestation.slot > state.validators[index].proof_of_custody_last_change else hash(state.validators[index].proof_of_custody_commitment)`. Append to `state.proof_of_custody_challenges` the object `ProofOfCustodyChallenge(responder_index=validator_index, seed_hash=seed_hash, depth=attestation.proof_of_custody_depth, data_root=attestation.data.shard_block_hash, data_index=data_index, expiry_slot=block.slot+PROOF_OF_CUSTODY_RESPONSE_DEADLINE, challenger_index=get_proposer(state, block), bit=bit)`.

Reviewer: A little Justinize:
```suggestion
* Let `seed_hash = state.validators[index].proof_of_custody_commitment if attestation.slot > state.validators[index].proof_of_custody_last_change else hash(state.validators[index].proof_of_custody_commitment)`.
* Append to `state.proof_of_custody_challenges` the object `ProofOfCustodyChallenge(responder_index=validator_index, seed_hash=seed_hash, depth=attestation.proof_of_custody_depth, data_root=attestation.data.shard_block_hash, data_index=data_index, expiry_slot=block.slot+PROOF_OF_CUSTODY_RESPONSE_DEADLINE, challenger_index=get_proposer(state, block), bit=bit)`.
```

A block can have maximum one proof of custody challenge, and it must appear before all `PROOF_OF_CUSTODY_SEED_CHANGE` objects.

#### PROOF_OF_CUSTODY_RESPONSE
```
{
    'challenge_index': 'uint64',
    'leaf': 'hash32',
    'root': 'hash32',
    'seed': 'hash32',
    'data_merkle_branch': '[hash32]',
    'poc_merkle_branch': '[hash32]',
}
```
Perform the following checks:

> * Verify `challenge_index < len(state.proof_of_custody_challenges)`.

Reviewer: Justinize:
```suggestion
* Verify that `challenge_index < len(state.proof_of_custody_challenges)`.
```

### Shard block fork choice rule

The fork choice rule for any shard is LMD GHOST using the validators currently assigned to that shard, but instead of being rooted in the genesis it is rooted in the block referenced in the most recent accepted crosslink (ie. `state.crosslinks[shard].shard_block_hash`). Only blocks whose `beacon_chain_ref` is the block in the main beacon chain at the specified `slot` should be considered (if the beacon chain skips a slot, then the block at that slot is considered to be the block in the beacon chain at the highest slot lower than a slot).

## Changes to beacon chain

Change to attestation verification:

> * Let `expected_depth = log2(SHARD_BLOCK_SIZE // SHARD_CHUNK_SIZE * next_power_of_2(slot - last_crosslink_slot))`. For each integer `i = 0, 1`, let `messages[i]` be `AttestationData(slot, shard, parent_hashes, shard_block_hash, expected_depth, attestation_indices.total_validator_count, justified_slot)`.

Reviewer 2: This is the only place `messages` is referenced in the entire phase 0 and phase 1 spec.
Seems like there is something missing here.

Also the index `i` is not used in the definition so `messages[0] == messages[1]`

> * Let `expected_depth = log2(SHARD_BLOCK_SIZE // SHARD_CHUNK_SIZE * next_power_of_2(slot - last_crosslink_slot))`. For each integer `i = 0, 1`, let `messages[i]` be `AttestationData(slot, shard, parent_hashes, shard_block_hash, expected_depth, attestation_indices.total_validator_count, justified_slot)`.

Reviewer 2: These fields don't match up with those defined in  `AttestationData`

A `ProofOfCustodyChallenge` has the following fields:

```python
{
    # Which validator is responding
    'responder_index': 'uint64',
    # Seed hash
    'seed_hash': 'hash32',

>     # Depth

Reviewer 3: Could you add a more detailed comment for this field? This comment doesn't add any additional information and `depth` is a bit vague w/o a deep understanding of the challenge itself.

    'depth': 'uint64',
    # Index in tree
    'data_index': 'uint64',
    # Expiry date
    'expiry_slot': 'uint64',
    # Who is challenging
    'challenger_index': 'uint64',
    # Data root
    'data_root': 'hash32',
    # Proof of custody bit
    'bit': 'bool'
}
```

> Three new types of `SpecialObject` records:

Reviewer 2: ```suggestion
Three additional types of `SpecialObject` records:
```

#### PROOF_OF_CUSTODY_SEED_CHANGE

```python
{
    'index': 'uint64',
    'new_commitment': 'hash32',

>     'signature': ['uint256']

Reviewer 2: change to `uint384`

```suggestion
    'signature': ['uint384']
```

}
```

> * Let `signed_data` be `bytes8(fork_version) + new_commitment`.

Reviewer 2: Need to utilize `BLSVerify` with `domain` rather than this old method

* Verify that `BLSVerify(pub=state.validators[index].pubkey, msg=hash(signed_data), sig=signature)` passes.

> * Verify that `hash(new_commitment) = state.validators[index].proof_of_custody_commitment`.

Reviewer 2: ```suggestion
* Verify that `hash(new_commitment) == state.validators[index].proof_of_custody_commitment`.
```

* Verify that `block.slot >= state.validators[index].proof_of_custody_last_change + PROOF_OF_CUSTODY_MIN_CHANGE_PERIOD`.

> * Set `state.validators[index].proof_of_custody_second_last_change` to `state.validators[index].proof_of_custody_last_change`.

Reviewer 2: Add line 
```
Set `state.validators[index].proof_of_custody_commitment` to `new_commitment`.
```

* Set `state.validators[index].proof_of_custody_last_change` to `block.slot`.

#### PROOF_OF_CUSTODY_CHALLENGE

```
{
    'attestation': SpecialAttestationData,
    'validator_index': 'uint64',
    'data_index': 'uint64',
}
```

Perform the following checks:

> * Verify that the attestation is valid.

Reviewer 2: ```suggestion
* Verify that the `attestation` is valid.
```

Reviewer 2: Note that you utilize `verify_special_attestation_data` for this validity check

* Verify that `attestation.slot > state.validators[index].proof_of_custody_second_last_change`.
* Verify that `validator_index in attestation.aggregate_sig_poc_0_indices` or `validator_index in attestation.aggregate_sig_poc_1_indices`; if the former, let `bit = False`, else `bit = True`.

> * Verify that `state.validators[index].status == ACTIVE` or `state.validators[index].status == PENDING_EXIT` and `block.slot < state.validators[index].exit_slot + PROOF_OF_CUSTODY_MIN_CHANGE_PERIOD`

Reviewer 2: Should a validator have to respond to PoC challenges if `EXITED` but not yet `WITHDRAWN`?
I suppose not. An `EXITED` validator shouldn't have to be online, and can only be punished for previously signed slashable messages.

> * Let `check_value = as_uint256(hash(as_bytes8(validator_index) + as_bytes8(attestation.slot) + as_bytes8(attestation.shard_id)))`.

Reviewer 2: This introduces big integer arithmetic. Been trying to avoid this requirement in eth2 clients outside of VM

> * Let `check_value = as_uint256(hash(as_bytes8(validator_index) + as_bytes8(attestation.slot) + as_bytes8(attestation.shard_id)))`.

Reviewer 2: What's going on there? Is this the mechanism to randomly reduce the requirement to store poc data in times of low validator count?

Can we not utilize a similar method with a 64 bit int? and just take the lower 8 bytes of the hash?

* Let `FULL_PROOF_OF_CUSTODY_VALCOUNT = TARGET_COMMITEE_SIZE * SHARD_COUNT`, and `target = 2**256 * attestation.total_validator_count // FULL_PROOF_OF_CUSTODY_VALCOUNT`.
* Verify that `check_value <= target`.

* Let `seed_hash = state.validators[index].proof_of_custody_commitment if attestation.slot > state.validators[index].proof_of_custody_last_change else hash(state.validators[index].proof_of_custody_commitment)`.

> * Append to `state.proof_of_custody_challenges` the object `ProofOfCustodyChallenge(responder_index=validator_index, seed_hash=seed_hash, depth=attestation.proof_of_custody_depth, data_root=attestation.data.shard_block_hash, data_index=data_index, expiry_slot=block.slot+PROOF_OF_CUSTODY_RESPONSE_DEADLINE, challenger_index=get_proposer(state, block), bit=bit)`.

Reviewer 2: `attestation` has no field `proof_of_custody_depth`

A block can have maximum one proof of custody challenge, and it must appear before all `PROOF_OF_CUSTODY_SEED_CHANGE` objects.

#### PROOF_OF_CUSTODY_RESPONSE
```
{
    'challenge_index': 'uint64',
    'leaf': 'hash32',
    'root': 'hash32',
    'seed': 'hash32',
    'data_merkle_branch': '[hash32]',
    'poc_merkle_branch': '[hash32]',
}
```
Perform the following checks:

* Verify that `challenge_index < len(state.proof_of_custody_challenges)`.
* Let`challenge = state.proof_of_custody_challenges[challenge_index]`.
* Verify that `hash(seed) = challenge.seed_hash`
* Verify that `verify_merkle_branch(leaf, data_merkle_branch, challenge.depth, challenge.data_index, root)` passes.
* Verify that `verify_merkle_branch(leaf, poc_merkle_branch, challenge.depth, challenge.data_index, xor(root, seed))` passes.

> * Verify that `uint256(root) % 2 == challenge.bit`.

Reviewer 2: does this count as big int arithmetic? Can we do a `&` with a `MASK` to pull out the bit instead?

> Remove `challenge` from `state.proof_of_custody_challenges` (note: if a block has multiple `PROOF_OF_CUSTODY_RESPONSE` objects , later objects' `challenge_index` values will need to take removals from earlier objects into account).

Reviewer 2: There's no way to clean up `challenges` if the validator doesn't respond. We need to add a mechanism to the `EPOCH` boundary calculations to go through the challenges, and remove/slash unresponded to ones that are past the deadline
